### ISSUE TITLE
Seems SemanticsObject leaks memory

### ISSUE THREAD

**Issue by xujim** [flutter/flutter#45599]
In my application, I need to setViewController for Engine manually, and found the app will crash when I turn on the voice over.
After some diggings, I suspect the semantics object leaks. It may not caused by flutter but iOS instead. See following memory graph, it is took when my app crashed:
![image](https://user-images.githubusercontent.com/3032016/69603715-67a01e00-1056-11ea-95fa-6c382fe3c1c9.png)
SemanticsObject is leaked because the __UIAccessibilityFocusedElements is holding him.
It crashed at accessing the bridge of SemanticsObject, because SetOwnerViewController for platformViewIOS will reset the accessibility bridge, it will update weak bridge for each semantics object when updating semantics, but for the leaked semantics object it failed to update and is nil as result:
![image](https://user-images.githubusercontent.com/3032016/69603875-d2e9f000-1056-11ea-932f-4caf311d2d8d.png)

I also attached the memory graph file:
[Runner[683].memgraph.zip](https://github.com/flutter/flutter/files/3890315/Runner.683.memgraph.zip)

## Steps to Reproduce
You can use following source to reproduce:
git: https://github.com/alibaba/flutter_boost.git
tag: 0.1.54
after download the source, open the example, and run. Then do following steps(click buttons) to reproduce:
```
push native -> push Flutter Page -> Open second route -> push native -> push Flutter Page -> Open second route -> push native -> back -> back-> back-> back-> back ..... crash
```
By the way, you can add codes to dump the object_ of AccessibilityBridge:
```objc
void AccessibilityBridge::clearState() {
    NSMutableString *log = [[NSMutableString alloc]initWithString:@"--dump objects before dealloc---\n"];
    for (NSNumber *uid in objects_.get()) {
        SemanticsObject *obj = objects_.get()[uid];
        NSString *lbl = [NSString stringWithCString:obj.node.label.c_str() encoding:NSUTF8StringEncoding];
        [log appendString:[NSString stringWithFormat:@"---%p-uid:%@-%p-label:%@-ref:%ld\n", obj, uid, obj.bridge.get(), lbl, obj.retainCount]];
        for(SemanticsObject *child in obj.children){
            [log appendString:[NSString stringWithFormat:@"---------%p-uid:%d--ref:%ld\n", child, child.uid, child.retainCount]];
        }
    }
    FML_DLOG(INFO) << "[XDEBUG]--AccessibilityBridge::clearState()---"<<[log UTF8String];
  [objects_ removeAllObjects];
  previous_route_id_ = 0;
  previous_routes_.clear();
}
```
here is the log in my side:
```
[VERBOSE0:accessibility_bridge.mm(790)] [XDEBUG]--AccessibilityBridge::clearState()-----dump objects before dealloc---
---0x10805c600-uid:0-0x2814fc5a0-label:-ref:1
---------0x10805c450-uid:1--ref:2
---0x10805c0f0-uid:3-0x2814fc5a0-label:Open second route-ref:3
---0x10805c2a0-uid:2-0x2814fc5a0-label:-ref:2
---------0x10804b010-uid:4--ref:2
---------0x10805c0f0-uid:3--ref:3
---0x10804ae60-uid:5-0x2814fc5a0-label:First Route-ref:2
---0x10805c450-uid:1-0x2814fc5a0-label:-ref:2
---------0x10805c2a0-uid:2--ref:2
---0x10804b010-uid:4-0x2814fc5a0-label:-ref:2
---------0x10804ae60-uid:5--ref:2
```
You can see the semantics object(address: 0x10805c0f0, uid:3)'s reference count is 3, but it should be 2. Other objects don't have the same problem.

**Target Platform:**
iOS
**Target OS version/browser:**
iOS 12.4.1, iPhoneX
**Devices:**
it occurs in physical devices

```
Flutter (Channel alf_stable_v1.5.4_json, v1.5.4-hotfix.4-pre.13, on Mac OS X
    10.14.6 18G95, locale zh-Hans-CN)
    • Flutter version 1.5.4-hotfix.4-pre.13 at
      /Users/yujie/Desktop/FlutterWork/flutter_gitlab_1.5
    • Framework revision c86f869fa6 (3 周前), 2019-11-03 21:12:54 +0800
    • Engine revision 52c7a1e849
    • Dart version 2.3.0 (build 2.3.0-dev.0.5 a1668566e5)

 
[!] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/yujie/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling
      support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor
      --android-licenses

[✓] iOS toolchain - develop for iOS devices (Xcode 10.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.2.1, Build version 10E1001
    • ios-deploy 1.9.4
    • CocoaPods version 1.8.3

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 39.0.3
    • Dart plugin version 191.8423
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)

[!] IntelliJ IDEA Ultimate Edition (version 2017.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] VS Code (version 1.40.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.6.0
```

**Comment by xujim** [comment]
It caused an issue in FlutterBoost: https://github.com/flutter/flutter/issues/45599

**Comment by xujim** [comment]
seems the focus will keep the object, is there a way to detach?

**Comment by iapicca** [comment]
Hi @xujim 
I see you are using an outdated version of flutter,
could you please run 
`
 flutter channel stable
`
`
 flutter upgrade
`
If the issue persist please provide your updated 
`flutter doctor -v`
thank you

**Comment by xujim** [comment]
hi, @iapicca the problem still exists: 
![image](https://user-images.githubusercontent.com/3032016/69625133-eca02d00-1080-11ea-8e99-d6d64097f034.png)
```

➜  ios git:(570f4ac) ✗ flutter doctor -v
[✓] Flutter (Channel stable, v1.9.1+hotfix.6, on Mac OS X 10.14.6 18G95, locale
    zh-Hans-CN)
    • Flutter version 1.9.1+hotfix.6 at
      /Users/yujie/Desktop/FlutterWork/flutter_github_1.9.1
    • Framework revision 68587a0916 (2 个月前), 2019-09-13 19:46:58 -0700
    • Engine revision b863200c37
    • Dart version 2.5.0

 
[!] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/yujie/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling
      support)
    • Platform android-29, build-tools 29.0.2
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor
      --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 10.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.2.1, Build version 10E1001
    • CocoaPods version 1.8.3

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 39.0.3
    • Dart plugin version 191.8423
    • Java version OpenJDK Runtime Environment (build
      1.8.0_202-release-1483-b49-5587405)

[!] IntelliJ IDEA Ultimate Edition (version 2017.2.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • For information about installing plugins, see
      https://flutter.dev/intellij-setup/#installing-the-plugins

[✓] VS Code (version 1.40.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.6.0

[✓] Connected device (1 available)
    • “yujie”的 iPhone • 7a4046657b77a617cf53b8a3b0ac97206f9f0df4 • ios • iOS
      12.4.1

! Doctor found issues in 2 categories.
```

**Comment by gaaclarke** [comment]
In comment https://github.com/flutter/engine/pull/14155#pullrequestreview-331422847 I recommend a fix for this issue:

> The correct fix is in the dealloc of the AccessibilityBridge. You need to either issue a command that gets __UIAccessibilityFocusedElements to forget about the object, or your need to nil out the bridge parameter of outstanding SemanticObjects.

**Comment by darrenaustin** [comment]
I tried to reproduce this problem by cloning the flutter_boost tag 0.1.54. However when I ran it the 'push native' button was not focusable with VoiceOver on so I was unable to reproduce that case you describe.

That said, it looks like a fix (https://github.com/flutter/engine/pull/13857) that I committed a while back would protect against this crash. Is it possible for you to test this against a version of the flutter engine that has that fix?

**Comment by zw1924** [comment]
> 我试图通过克隆flutter_boost标签0.1.54重现此问题。但是，当我运行它时，“推动本机”按钮无法在VoiceOver上打开，因此我无法重现您描述的情况。
> 
> 就是说，我[不久前](https://github.com/flutter/engine/pull/13857)提交的修复程序（[flutter / engine＃13857](https://github.com/flutter/engine/pull/13857)）可以防止崩溃。您是否可以针对具有该修复程序的Flutter引擎版本进行测试？

I referenced your changes, and reworked the engine. Unfortunately, on iOS 13.* later, turning on the auxiliary function -> reading screen will trigger the crash exactly.
`
std::__1::unique_ptr<flutter::AccessibilityBridge, std::__1::default_delete<flutter::AccessibilityBridge> >::reset(flutter::AccessibilityBridge*) + 76
`

**Comment by jxw1102** [comment]
I also have a memory leak when I use webview_flutter plugin.

I got the information below with "Debug Memory Graph" of Xcode, i shows a retain cycle:
**FlutterSemanticsObject** has FlutterPlatformViewSemanticsContainer; 
FlutterPlatformViewSemanticsContainer has accessibilityElements;
accessibilityElements contains **FlutterSemanticsObject** and FLTWKWebView.

```
(lldb) po [[0x7f836fe5eee0 platformViewSemanticsContainer] accessibilityElements]
<__NSArrayI 0x6000026d5d60>(
<FlutterSemanticsObject: 0x7f836fe5eee0>,
<FLTWKWebView: 0x7f8370916400; frame = (0 0; 375 712); autoresize = W+H; layer = <CALayer: 0x6000026d2500>>
)
```

**Comment by zoeyfan** [comment]
Spoke with the customer and clarified that it is a blocker for them. In addition, they submitted a PR here: https://github.com/flutter/engine/pull/14155

**Comment by gaaclarke** [comment]
https://github.com/flutter/engine/pull/17499 May have fixed this.

**Comment by xster** [comment]
@xujim can you try again on master?

**Comment by Hixie** [comment]
@gaaclarke assigned this to you since we believe you fixed it, in case it gets reopened.
@xujim please let us know if this is not fixed. For now we will assume that @gaaclarke's PR fixed it.
Thanks!

**Comment by gaaclarke** [comment]
I tried to reproduce tag 1.5.4 of the aforementioned repository.  It doesn't compile (`return`ing void value), when I fix the compilation errors it crashes when it launches:

```
[✓] Flutter (Channel master, v1.18.0-6.0.pre.35, on Mac OS X 10.14.6 18G2022, locale en-US)
    • Flutter version 1.18.0-6.0.pre.35 at /Users/aaclarke/dev/flutter
    • Framework revision c97f61a452 (28 minutes ago), 2020-04-15 13:12:14 -0700
    • Engine revision deef2663ac
    • Dart version 2.8.0 (build 2.8.0-dev.20.0 c9710e5059)
```

**Comment by gaaclarke** [comment]
Further evidence that the root problem is addressed:  I pushed and popped a FlutterViewController with VoiceOver active, did a "Debug Memory Graph" and there are no live `FlutterSemanticesObject`s that leaked.

**Comment by zoeyfan** [comment]
@xujim can you please confirm this is no longer an issue for you?
@kangwang1988 FYI.

**Comment by xster** [comment]
Tentatively closing this issue since we have some degree of confidence that it's since been fixed. Please reopen if incorrect @xujim

**Comment by alexmarkley** [comment]
I'm leaving this comment for posterity because this issue was killing me, and the connection to the webview_flutter plugin was not completely obvious.

On current stable (1.17.5) my project is leaking WKWebViews all over the place, whereas in beta 1.19.0-4.3.pre the problem appears to be gone.

So if you are using webview_flutter w/flutter stable and your HTML5 video continues to play even after you have popped the WebView widget, try the latest beta version of flutter.

Thanks all for working on this!
